### PR TITLE
Resolve vulnerabilities in azul-elasticsearch image (DataBiosphere/azul-private#93

### PR DESCRIPTION
**Pull request by dsotirho-ucsc**
Connected issues: DataBiosphere/azul-private#93

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] PR and connected issue are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] PR has checklist items for upgrading instructions <sub>or PR is not labeled `upgrade`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Manually deployed `dev.shared`
- [x] Confirmed that the previous, now unused, image version was deleted from ECR in `dev`
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Manually deployed `anvildev.shared`
- [x] Confirmed that the previous, now unused, image version was deleted from ECR in `anvildev`
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Manually deployed `anvilprod.shared`
- [x] Confirmed that the previous, now unused, image version was deleted from ECR in `anvilprod`
- [x] Pushed PR branch to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `hammerbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Added commit title tags to merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Build passes on GitLab `anvilprod`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilprod` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvilprod` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Added CL item to `prod` promotion PR:
> - [ ] Manually deployed `prod.shared`
- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by dsotirho-ucsc** [comment]
Tested PR by deploying to `dev.shared` and ran all unit tests locally confirming the new `7.17.10-3` image was being used.

```
...
2023-09-14 03:26:41,595 INFO    MainThread: Launching container from image 122796619775.dkr.ecr.us-east-1.amazonaws.com/docker.io/ucscgi/azul-elasticsearch:7.17.10-3
2023-09-14 03:26:42,210 INFO    MainThread: Launched container gallant_lederberg from image 122796619775.dkr.ecr.us-east-1.amazonaws.com/docker.io/ucscgi/azul-elasticsearch:7.17.10-3, with container port 9200 mapped to 127.0.0.1:52657 on the host
2023-09-14 03:26:42,211 DEBUG   MainThread: Creating ES client [127.0.0.1:52657]
...
```

**Review by nadove-ucsc — DISMISSED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Conflicts. Looks good otherwise.

**Review by hannes-ucsc — DISMISSED** [review]

**Comment by hannes-ucsc** [comment]
No tags in PR titles, please.

File: UPGRADING.rst [hunk 1]
```diff
@@ -20,6 +20,22 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
...
 
+DataBiosphere/azul-private#93 Resolve vulnerabilities in azul-elasticsearch
+===========================================================================
+
+Operator
+~~~~~~~~
+
+Before pushing the PR branch to the ``sandbox``, ``anvilbox``, or ``hammerbox``
+deployment, manually deploy the ``shared`` component of corresponding main
+deployment. If the PR fails during testing and is not merged, rollback the
```
Comment (nadove-ucsc) [inline UPGRADING.rst]: "rollback" is a noun, "roll back" is the verb form.

```suggestion
deployment. If the PR fails during testing and is not merged, roll back the
```

File: UPGRADING.rst [hunk 2]
```diff
@@ -20,6 +20,22 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
...
 
 
+DataBiosphere/azul-private#93 Resolve vulnerabilities in azul-elasticsearch
+===========================================================================
+
+Operator
+~~~~~~~~
+
+Before pushing the PR branch to the ``sandbox``, ``anvilbox``, or ``hammerbox``
+deployment, manually deploy the ``shared`` component of corresponding main
```
Comment (nadove-ucsc) [inline UPGRADING.rst]:
```suggestion
deployments, manually deploy the ``shared`` component of the corresponding main
```

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]
Just some nitpicks of the upgrading prose.

**Review by nadove-ucsc — DISMISSED** [review]

**Review by hannes-ucsc — DISMISSED** [review]

**Comment by dsotirho-ucsc** [comment]
Re-requesting review, previous approval was auto-dismissed after adding `requirements_update` commit.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
No need to update requirements in this PR.

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Conflicts.

File: src/azul/__init__.py [hunk 3]
```diff
@@ -1432,7 +1432,7 @@ def docker_registry(self) -> str:
     # the `shared` TF component.
 
     docker_images = [
-        'docker.io/ucscgi/azul-elasticsearch:7.17.10-2',
+        'docker.io/ucscgi/azul-elasticsearch:7.17.10-3',
```
Comment (hannes-ucsc) [inline src/azul/__init__.py]: Please ask @achave11-ucsc to bump the internal version (he can push to the `main` branch now, without a PR) and then 

```suggestion
        'docker.io/ucscgi/azul-elasticsearch:7.17.10-4,
```

File: test/es_test_case.py [hunk 4]
```diff
@@ -36,7 +36,7 @@ class ElasticsearchTestCase(DockerContainerTestCase):
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
-        es_endpoint = cls._create_container('docker.io/ucscgi/azul-elasticsearch:7.17.10-2',
+        es_endpoint = cls._create_container('docker.io/ucscgi/azul-elasticsearch:7.17.10-3',
```
Comment (hannes-ucsc) [inline test/es_test_case.py]:
```suggestion
        es_endpoint = cls._create_container('docker.io/ucscgi/azul-elasticsearch:7.17.10-4',
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/__init__.py (hunk 3 above)
Comment (achave11-ucsc) [reply src/azul/__init__.py]: I'm unable to push to the `main` branch. I get the following error:
```
❯ git push github main
Enumerating objects: 9, done.
Counting objects: 100% (9/9), done.
Delta compression using up to 10 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (5/5), 465 bytes | 465.00 KiB/s, done.
Total 5 (delta 1), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
remote: error: GH006: Protected branch update failed for refs/heads/main.
remote: error: Required status check "build" is expected.
To https://github.com/DataBiosphere/azul-docker-elasticsearch.git
 ! [remote rejected] main -> main (protected branch hook declined)
error: failed to push some refs to 'https://github.com/DataBiosphere/azul-docker-elasticsearch.git'
```

Should I just push the change on a separated branch, get the build check, and then commit on `main`?

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**